oak_remote_attestation/src/handler.rs

```diff
@@ -70,7 +73,54 @@ impl<H: FnOnce(Vec<u8>) -> Vec<u8>> EncryptionHandler<H> {
 
         // Handle request.
         let response = (self.request_handler)(request);
-        log::info!("plaintext response: {:?}", response);
+
+        // Encrypt and serialize response.
+        // The resulting decryptor for consequent requests is discarded because we don't expect
+        // another message from the stream.
+        server_encryptor
+            .encrypt(&response, EMPTY_ASSOCIATED_DATA)
+            .context("couldn't encrypt response")
+    }
+}
+
+/// Wraps a closure to an underlying function with request encryption and response decryption logic,
+/// based on the provided encryption key.
+/// `AsyncEncryptionHandler` can be used when the `AsyncRecipientContextGenerator` is needed.
+pub struct AsyncEncryptionHandler<H: FnOnce(Vec<u8>) -> Vec<u8>> {
+    // TODO(#3442): Use attester to attest to the public key.
+    recipient_context_generator: Arc<dyn AsyncRecipientContextGenerator + Send + Sync>,
+    request_handler: H,
+}
+
+impl<H: FnOnce(Vec<u8>) -> Vec<u8>> AsyncEncryptionHandler<H> {
+    pub fn create(
+        recipient_context_generator: Arc<dyn AsyncRecipientContextGenerator + Send + Sync>,
+        request_handler: H,
+    ) -> Self {
+        Self {
+            recipient_context_generator,
+            request_handler,
+        }
+    }
+}
+
+impl<H: FnOnce(Vec<u8>) -> Vec<u8>> AsyncEncryptionHandler<H> {
+    pub async fn invoke(
+        self,
+        encrypted_request: &EncryptedRequest,
+    ) -> anyhow::Result<EncryptedResponse> {
+        // Initialize server encryptor.
+        let mut server_encryptor =
+            AsyncServerEncryptor::new(self.recipient_context_generator.as_ref());
+
+        // Decrypt request.
+        let (request, _) = server_encryptor
```

Comment:
Is the second value the associated data? Maybe leave it in as `_associated_data` (or its proper name) for clarity